### ISSUE TITLE
[proposal] Support for machine/board Bundles

### ISSUE THREAD

**Issue by ya-mouse** [tinygo-org/tinygo#3362]
Within a time we will have a significant number of supported machines/boards. This will bloat the base tinygo srcs.

**Proposal**

This proposal is about to split the machines-related code in "bundles". The existing code allows to merge regular Go sources and TinyGo (btw, src/os/exec/exec.go never being used with current implementation because os/exec is overriden by Go sources). The approach is to traverse bundles `src` directories and merge the contents with the TinyGo sources in the cached root. Same should be done for `lib` and `bin` if present. The per bundle `targets` should be merged separately. 

**Dependencies**
In order to support cross-bundle dependencies, it's possible to wrap the `go install` command to be called within bundle directory (set the `GOPATH` variable). The existing PoC should be re-worked to follow pkg-structure of Go packages.

**Caching**
At the first implementation the `targets` being traversed each invocation and only missing symlinks are re-created.
The bundle modifications (install, update, remove) should invalidate the existing cache.

The PoC that splits up the `nrfx` package: https://github.com/tinygo-org/tinygo/pull/3361

**Comment by aykevl** [comment]
This is a pretty drastic change. What's the advantage of this new system? Is there an advantage other than reducing the number of files in the machine package?

**Comment by ya-mouse** [comment]
I will split the change. The main part is for actual bundles support and the rest is files bouncing around.
The advantage to move out machine/boards/targets out of main src tree.

**Comment by aykevl** [comment]
> The advantage to move out machine/boards/targets out of main src tree.

Ah I see.  In that case, I think the easier option would be to move the machine package as a whole out of the tinygo tree. This is something I've wanted to do for a while but it is rather difficult to do and would probably break backwards compatibility with existing code.

**Comment by ya-mouse** [comment]
The proposed change and working proof of concept do seamless merge of files for machine package and runtime as well. The bunch of files for machines are also there. With no backward compatibility breakage.

**Comment by aykevl** [comment]
> The proposed change and working proof of concept do seamless merge of files for machine package and runtime as well.

I would very much prefer to _not_ do that. We already do some merging of the Go standard library. We already have two variants (one with the syscall package and one without), and these variants are already troublesome for things like IDE support. Please don't introduce more variants for chip families.

**Comment by ya-mouse** [comment]
So, the only way is to move into separate packages out of the main tree? Then, the changes will not be back compatible. And things like uart from existing machine/runtime will require some extra init function to pass the machine's uart instance and so on.

**Comment by digitalentity** [comment]
I ran into this recently - solved by keeping tinygo toolchain downloaded separately and symlinking additional files into the `src/machine` directory. I think this approach can be generalized:

## Precedent

Target specifications already solve the analogous problem for build metadata: `-target=./custom.json` loads an out-of-tree target spec, which may inherit from built-in targets. This proposal extends the same mechanism to `machine` package sources.

## Design

Two new fields in the target JSON specification:

```json
{
    "inherits": ["rp2350"],
    "build-tags": ["myboard"],
    "machine-dir": "/custom/path/to/machine",
    "machine-extra-dirs": ["/overlay/path/for/board/package"]
}
```

### `machine-dir` (string)

Wholesale replacement. The base for `src/machine` in the synthetic GOROOT becomes this directory instead of `$TINYGOROOT/src/machine`. The replaced package must be a complete `machine` implementation for the selected build tags; users may vendor or symlink upstream files they want to keep. Defaults to `$TINYGOROOT/src/machine` if unset.

### `machine-extra-dirs` ([]string)

Additive overlay. Regular files from each listed directory are placed on top of the base machine directory (the `machine-dir` if set, otherwise `$TINYGOROOT/src/machine`). This is the cheap path for the common case: a custom board on a stock chip needs one `board_myboard.go` file, gated by a build tag declared in the same JSON.

## Why this fits the existing architecture

- **Import path is unchanged.** The package is still imported as `"machine"`,  so the driver ecosystem, the `runtime` package coupling, and the compiler  intrinsics keyed on `machine.*` symbol names (`compiler/intrinsics.go`,  `compiler/symbol.go`) are all untouched.
- **Overlay files are full citizens.** Because they are merged into the synthetic GOROOT, they may import `device/*` and be imported by `runtime`, with no linkname tricks required.
- **Caching is correct for free.** The synthetic GOROOT cache key is a hash of the computed merge-link map (`loader/goroot.go`), which contains the source  paths of every link. A different override directory or a changed file set  produces a different map, hence a different cached GOROOT. Content edits are  picked up live through the symlinks.
- **File selection is unchanged.** Which files compile is still decided by build tags from the target spec, exactly as for in-tree machine files.

This also unblocks the possibility to gradually split the existing `src/machine` into separate directories by e.g. the MCU family (maybe even submodules for simplicity).

@deadprogram @aykevl what do you think?

**Comment by deadprogram** [comment]
I did some experimenting with symlinking a while back https://github.com/deadprogram/tinygo-external-board

**Comment by digitalentity** [comment]
> I did some experimenting with symlinking a while back https://github.com/deadprogram/tinygo-external-board

This is pretty much the same think I did for my project.

I thought about making `machine` a separate package altogether (e.g. `tinygo.org/x/stm32/machine`), but this breaks the fundamental Go assumption about module import path. Some code says `import "machine"` - that path resolves only inside GOROOT. Module tinygo.org/x/stm32/machine with package machine inside must be imported as tinygo.org/x/stm32/machine, unless we want to diverge from Go spec and introduce module aliasing so `machine` may actually resolve outside of the `GOROOT`. Not impossible and TBH I would prefer exactly this approach.

**Comment by digitalentity** [comment]
Actually...

How about this:

```json
{
    "inherits": ["rp2350"],
    "build-tags": ["myboard"],
    "machine": "tinygo.org/x/stm32-machine@v1.2.3"
}
```

The `machine` (string)

Wholesale replacement from a versioned Go module: `module/path@version`. TinyGo resolves the module to a directory (`go mod download -json`) and uses that directory symlink'd into the `src/machine`. This makes a vendor's `machine` implementation a fetchable, versioned artifact instead of a compiler fork, while the import path - and therefore type identity, driver
compatibility, the `runtime` coupling, and the `machine.*` compiler intrinsics - remain untouched.

**Constraint**: the module's package is compiled as a GOROOT package, so it may import only GOROOT packages (`device/*`, `runtime/volatile`, `internal/binary`, `unsafe`, ...) but **may not depend on other Go modules**.
